src/sinks/opendal_common.rs

```diff
@@ -0,0 +1,218 @@
+//! opendal_common provide real sink supports for all opendal based services.
+//!
+//! # TODO
+//!
+//! opendal service now only support very basic sink features. To make it
+//! useful, we need to add the following features:
+//!
+//! - Batch events
+//! - Error handling
+//! - Limitation
+//! - Compression
+//! - KeyPartition
+
+use crate::event::EventFinalizers;
+use crate::sinks::util::encoding::write_all;
+use crate::sinks::util::encoding::Encoder;
+use crate::sinks::util::metadata::RequestMetadataBuilder;
+use crate::sinks::util::{request_builder::EncodeResult, Compression};
+use crate::sinks::BoxFuture;
+use crate::{
+    event::Event,
+    internal_events::SinkRequestBuildError,
+    sinks::util::{RequestBuilder, SinkBuilderExt},
+};
+use bytes::Bytes;
+use futures::{stream::BoxStream, StreamExt};
+use opendal::Operator;
+use snafu::Snafu;
+use std::task::Poll;
+use tower::Service;
+use vector_common::finalization::{EventStatus, Finalizable};
+use vector_common::request_metadata::MetaDescriptive;
+use vector_common::request_metadata::RequestMetadata;
+use vector_core::internal_event::CountByteSize;
+use vector_core::sink::StreamSink;
+use vector_core::stream::DriverResponse;
+
+pub struct OpendalSink {
+    op: Operator,
+}
+
+impl OpendalSink {
+    pub fn new(op: Operator) -> Self {
+        Self { op }
+    }
+}
+
+#[async_trait::async_trait]
+impl StreamSink<Event> for OpendalSink {
+    async fn run(
+        self: Box<Self>,
+        input: futures_util::stream::BoxStream<'_, Event>,
+    ) -> Result<(), ()> {
+        self.run_inner(input).await
+    }
+}
+
+impl OpendalSink {
+    async fn run_inner(self: Box<Self>, input: BoxStream<'_, Event>) -> Result<(), ()> {
+        input
+            .request_builder(
+                None,
+                OpendalRequestBuilder {
+                    encoder: OpendalEncoder,
+                },
+            )
+            .filter_map(|request| async move {
+                match request {
+                    Err(error) => {
+                        emit!(SinkRequestBuildError { error });
+                        None
+                    }
+                    Ok(req) => Some(req),
+                }
+            })
+            .into_driver(OpendalService::new(self.op.clone()))
+            .run()
+            .await
+    }
+}
+
+/// TODO: we should implment batch encoder.
+#[derive(Clone)]
+struct OpendalEncoder;
+
+impl Encoder<Event> for OpendalEncoder {
+    fn encode_input(
+        &self,
+        input: Event,
+        writer: &mut dyn std::io::Write,
+    ) -> std::io::Result<usize> {
+        let event = serde_json::to_string(&input).unwrap();
+        write_all(writer, 1, event.as_bytes()).map(|()| event.len())
+    }
+}
```

Comment:
**thought (non-blocking)**: It'd be nice to tie this back into the [codecs](https://rust-doc.vector.dev/vector/codecs/index.html) crate for better re-usability.